### ISSUE TITLE
How to match with date fields?

### ISSUE THREAD

**Issue by Mike-Cleaver-LB** [SalesforceLabs/Configurable-Self-Registration-for-Experience-Cloud#40]
Hi team, I am trying to add DOB as a date field and use it for matching.
I have set up a custom DOB field on the user object, and the query references it (several times, the client wants a 3 of 5 match)
but the error is "`Invalid bind expression type of String for column of type Date`" It is passing the date yyyy-mm-dd into the query successfully by my understanding, but presumably it is in the wrong format? I don't see anyway to set the format on a date input

I know how to do it with some apex, but I don't have that option here. Any ideas would be welcome!

The log message is as follows

Form Values: `{Birthdate=1987-12-10, Email=[mike.cleaver+dobtest@.....[trimmed].... password=HIDDEN}`
There was a problem executing the specific query in the Custom Query property. Query used: 
```
SELECT id FROM Account WHERE (firstname = :firstName AND lastname = :lastName AND personemail = :email) or...[trimmed]... or(firstname = :firstName AND lastname = :lastName and PersonBirthdate = :Birthdate) or ...[trimmed]... or (personemail = :email AND phone = :MobilePhone and PersonBirthdate = :Birthdate).

Error: Invalid bind expression type of String for column of type Date
```

**Comment by jlowe-SFDC** [comment]
Hi @Mike-Cleaver-LB - This should work fine. I'm using a date match for my client. How is the date field configured? It is a date on the user object and in the form?

My example so you can compare configuration:

Log record:
Searching for record with query:SELECT Id,AccountId FROM Case WHERE Crime_Ref_Number__c = :Crime_Ref_Number__c AND Contact_Date_of_Birth__c = :Contact_Date_of_Birth__c AND (ContactEmail = :Email OR Mobile_Phone_Unformatted__c = :Mobile_Phone_Unformatted__c) LIMIT 1. 

Submitted Values: {Contact_Date_of_Birth__c=2001-01-01, Crime_Ref_Number__c=21240000219, Email=[imtest.tc20@gmail.com](mailto:imtest.tc20@gmail.com), FirstName=JACK, LastName=TEST, Mobile_Phone_Unformatted__c=07448411354, confirmPassword=HIDDEN, password=HIDDEN}

Form setup:
![image](https://github.com/user-attachments/assets/983a027b-0228-4315-a20b-edf76d6f40f6)

Field in SF:
![image](https://github.com/user-attachments/assets/7bb82131-8886-41e9-911b-8da3fcf612a1)

Let me know and might be able to identify if there is an issue to fix.

**Comment by Mike-Cleaver-LB** [comment]
So the value I'm comparing to in the record is a date 

it's called personBirtdate on the account object (we're using person accounts)
![image](https://github.com/user-attachments/assets/8edb5526-486f-41dd-aac9-d9286f18e944)

it's called birthdate on the user record
![image](https://github.com/user-attachments/assets/fc792b81-b89c-49b9-952d-6b10e8edd04c)

![image](https://github.com/user-attachments/assets/6e6a5545-7150-4231-b4f6-d261df7f8a58)
is how the metadata is setup for this field.

I might just temporarily to one check on date only and see if that behaves differently

**Comment by Mike-Cleaver-LB** [comment]
I corrected my metadata to be the user api name for birthdate and the error is similar

...
Username valid.
An error occurred whilst trying to update the User Object:Illegal assignment from String to Date
Form Inputs: {Birthdate__c=1987-12-10,

**Comment by Mike-Cleaver-LB** [comment]
I had a look at the changes, it seems that you're using convertFormDataToCorrectDataType to parse the data, but it doesn't seem to handle strings? it should also be a 'switch on' rather than cascading if statements, it will only be one, no need to keep matching on the remaining ifs when you have parsed the value already?

`Schema.DisplayType dataType = objectMetadata.get(key).getDescribe().getType();`
becomes
```
string dataType = SiteUtilities.convertToString( objectMetadata.get(key).getDescribe().getType() );

switch on (datatype){
when 'DATE' {
}
when....{
}
when else {
// handle all different basic strings, like multiselect or whatever here
}
```

as I read it, basic strings won't get added to the list anymore?

**Comment by Mike-Cleaver-LB** [comment]
I tried creating a PR for this, I dunno if it worked, i got an error about the CLA which I signed already. Are you able to see it from your side?

**Comment by Mike-Cleaver-LB** [comment]
Oh damn, old habits sorry, I used to use Java before moving to SF a few years ago. Ok, that's disappointing, I will update my code and send it through.

And i see that you're using the passed in map to do the transforming. I didn't realise that, again from learned history I don't tend to manipulate the values I am going through. But you're right, we only need to modify the values we are actually  changing. I can update my code to reflect that.

PR updated/sent and Linkdin request sent!

**Comment by jlowe-SFDC** [comment]
This is turning out to be more complex than originally planned! 

Where I removed the checkFieldExists() check at multiple points, and had it in the convertFormDataToCorrectType() method, it was removing the 'Email' field from the submitted values when using Person Accounts (which was expected). The issue with this is that the Custom Query then doesn't work because it might expect :Email as part of the query and then fails.

This works fine, as long as the user configures custom metadata appropriately i.e. when setting things up and using Person Accounts, create a new entry for PersonEmail and disable the included out the box one. Ensure that any query references the correct value and you're good to go. The issue is that this is a change in behaviour and may break orgs where they upgrade.

Therefore, I've reverted this part of the change and removed the checkFieldExists from convertFormDataToCorrectType() instead.
